### PR TITLE
Refactor EstimateSwap (https://github.com/osmosis-labs/osmosis/pull/2378)

### PR DESCRIPTION
**Pull request by mattverse**
Sub component of  https://github.com/osmosis-labs/osmosis/pull/2378

## What is the purpose of the change

https://github.com/osmosis-labs/osmosis/pull/2378 has some parts in code where code de-duplication can take place. This PR includes minor refactoring to for code cleanup and code de-duplication.

## Brief Changelog
- Remove code duplication for Estimate methods

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Comment by czarcas7ic** [comment]
It seems the estimate_swap.go file isn't covered by any tests. Is there a plan to add this in?

**Comment by mattverse** [comment]
Good point! I'll be adding them on this PR

**Comment by ValarDragon** [comment]
This is missing the proto field removal right?

**Comment by vuong177** [comment]
> This is missing the proto field removal right

Target branch has it.

**Comment by mattverse** [comment]
@ValarDragon yeah this is PRd against the target branch not the main branch. Only contains refactored logic

File: x/gamm/keeper/estimate_swap.go [hunk 1]
```diff
@@ -158,36 +154,32 @@ func (k Keeper) estimateSwapExactAmountOut(
...
 
-	err = k.EstimateUpdatePoolForSwap(ctx, pool, tokenIn, tokenOut)
-	if err != nil {
-		return sdk.Int{}, err
-	}
-	return tokenInAmount, nil
+	return pool, tokenIn, nil
 }
 
 func (k Keeper) EstimateUpdatePoolForSwap(
```
Comment (vuong177) [inline x/gamm/keeper/estimate_swap.go:185]: should we remove `EstimateUpdatePoolForSwap` func ?

Comment (mattverse) [reply x/gamm/keeper/estimate_swap.go:185]: good catch, its not being used anywhere anymore, removed

File: wasmbinding/test/messages_test.go [hunk 2]
```diff
@@ -636,6 +636,8 @@ func TestSwap(t *testing.T) {
 	for name, spec := range specs {
 		t.Run(name, func(t *testing.T) {
 			// when
+			fmt.Println("=====swap message")
```
[the remark this answers is not in the document]
Comment (mattverse) [reply wasmbinding/test/messages_test.go]: done!

**Comment by mattverse** [comment]
@vuong177 re-requested for a second round of review, please feel free to drop more reviews or comments, otherwise would like to merge this to you branch when I have your approval on this PR

File: x/gamm/keeper/estimate_swap.go [hunk 3]
```diff
@@ -148,7 +143,8 @@ func (k Keeper) EstimateSwapExactAmountOut(
 		return sdk.Int{}, err
 	}
 	swapFee := pool.GetSwapFee(ctx)
-	return k.estimateSwapExactAmountOut(ctx, pool, tokenInDenom, tokenInMaxAmount, tokenOut, swapFee)
+	_, tokenIn, err := k.estimateSwapExactAmountOut(ctx, pool, tokenInDenom, tokenInMaxAmount, tokenOut, swapFee)
+	return tokenIn.Amount, err
```
Comment (vuong177) [inline x/gamm/keeper/estimate_swap.go:147]:
```suggestion
	if err != nil {
		return tokenOut.Amount, err
	}

	err = k.EstimateUpdatePoolForSwap(ctx, pool, tokenIn, tokenOut)
	return tokenOut.Amount, err
```

File: x/gamm/keeper/estimate_swap.go [hunk 4]
```diff
@@ -94,7 +94,8 @@ func (k Keeper) EstimateSwapExactAmountIn(
 	}
 
 	swapFee := pool.GetSwapFee(ctx)
-	return k.estimateSwapExactAmountIn(ctx, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
+	_, tokenOut, err := k.estimateSwapExactAmountIn(ctx, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
+	return tokenOut.Amount, err
```
Comment (vuong177) [inline x/gamm/keeper/estimate_swap.go:98]:
```suggestion
	if err != nil {
		return tokenOut.Amount, err
	}

	err = k.EstimateUpdatePoolForSwap(ctx, pool, tokenIn, tokenOut)
	return tokenOut.Amount, err
```

File: x/gamm/keeper/estimate_swap.go [hunk 5]
```diff
@@ -158,54 +154,30 @@ func (k Keeper) estimateSwapExactAmountOut(
...
-	tokensIn := sdk.Coins{tokenIn}
-	tokensOut := sdk.Coins{tokenOut}
-
-	err := k.setPool(ctx, pool)
-	if err != nil {
-		return err
-	}
-
-	k.RecordTotalLiquidityIncrease(ctx, tokensIn)
-	k.RecordTotalLiquidityDecrease(ctx, tokensOut)
```
Comment (vuong177) [inline x/gamm/keeper/estimate_swap.go:208]: revert `EstimateUpdatePoolForSwap`

**Comment by vuong177** [comment]
hey @mattverse, I think we need to revert `EstimateUpdatePoolForSwap`. I've create a test for estimate swap, and seems it's failed because of missing update pool when using multi-route in swap.

**Comment by mattverse** [comment]
@vuong177 I'm not sure if I follow, I thought one of the goals of changing `EstimateSwap` was to change so that 1) remove the process of actual swapping 2) remove updating pool upon querying estimate swap. IMO, changing the state of the pool after estimating a swap feels a little weird. wdyt?

**Comment by vuong177** [comment]
@mattverse, yes but 

> @vuong177 I'm not sure if I follow, I thought one of the goals of changing `EstimateSwap` was to change so that 1) remove the process of actual swapping 2) remove updating pool upon querying estimate swap. IMO, changing the state of the pool after estimating a swap feels a little weird. wdyt?

yes, but if not update pool, we'll get the wrong result when estimate swap with multi-route (you can see my test).

**Comment by vuong177** [comment]
I see we use `CacheContext` on in `EstimateMultihopSwapExactAmountOut` and `EstimateMultihopSwapExactAmountIn`, is it safe?
```
        sdkCtx := sdk.UnwrapSDKContext(ctx)
	subCtx, _ := sdkCtx.CacheContext()

	tokenOutAmount, err := q.Keeper.EstimateMultihopSwapExactAmountIn(subCtx, req.Routes, tokenIn, sdk.NewInt(1))
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}
````

**Comment by mattverse** [comment]
@vuong177 Ahh I see what you mean, you're saying that we're doing a multi-hop thus need a changed state from the previous multihop

**Comment by mattverse** [comment]
@vuong177  Reverted `EstimateUpdatePoolForSwap`, tests are still failing for some reason :( Would you mind taking over this branch to fix so that the test passes?

**Comment by vuong177** [comment]
@mattverse sure, I'm fixing it.

**Comment by vuong177** [comment]
Should I merge it to the old branch, and create a PR from it to master  @mattverse ?

**Comment by mattverse** [comment]
@vuong177 Yeah! I'm on that rn! 

We decided to make this into V12, so I want to push this fix forward asap

**Comment by mattverse** [comment]
Proto check failing rn, but merging this regardless as this is not a merge to main, but just a different branch for the PR, planning to fix proto checks and merge conflicts off that branch

**Pull request merged**